src/builders/thermodynamic.jl

```diff
@@ -0,0 +1,258 @@
+    water_ids = ["h2o_c", "h2o_e"],
+    concentration_lb = 1e-9,
+    concentration_ub = 100e-3,
+    T = 298.15, # Kelvin
+    R = 8.31446261815324e-3, # kJ/K/mol
+    ignore_reaction_ids = String[],
+)
+    # check if reactions that will be used in the model all have thermodynamic data, otherwise throw error.
+    # only these reactions will form part of the model.
+    rxns = filter(
+        x -> !(x in ignore_reaction_ids),
+        isempty(flux_solution) ? A.reactions(model) : collect(keys(flux_solution)),
+    )
+    all(in.(rxns, Ref(collect(keys(reaction_standard_gibbs_free_energies))))) || throw(
+        ArgumentError("""
+                      Not all reactions have thermodynamic data. 
+                      Either add the reactions with missing ΔG0s to ignore_reaction_ids, 
+                      or add the data to reaction_standard_gibbs_free_energies.
+                      """),
+    )
+
+    # import metabolite ids (if used), and reaction stoichiometries from an AbstractFBCModel
+    mets = Set(Symbol[])
+    for rid in rxns
+        for met in keys(A.reaction_stoichiometry(model, rid))
+            push!(mets, Symbol(met))
+        end
+    end
+    mets = [m for m in mets] # TODO: constrainttrees #12
+    stoi = A.stoichiometry(model)
+
+    # create thermodynamic variables
+    m = C.ConstraintTree(:max_min_driving_force^C.variable())
+    m +=
+        :log_metabolite_concentrations^C.variables(
+            keys = mets,
+            bounds = zip(log(concentration_lb), log(concentration_ub)),
+        )
+    m += :delta_G_reactions^C.variables(keys = Symbol.(rxns))
+
+    #= 
+    Build gibbs free energy relations
+    ΔG_rxn == ΔG0 + R * T * sum ( log_concentration_variable * stoichiometry_value )
+    =#
+    model_rxns = A.reactions(model)
+    model_mets = A.metabolites(model)
+    for rxn in rxns
+        met_idxs, stoich_coeffs = findnz(stoi[:, findfirst(==(rxn), model_rxns)])
+        met_ids = model_mets[met_idxs]
+
+        dG0 = reaction_standard_gibbs_free_energies[rxn]
+
+        m *=
+            :delta_G_reaction_equations^Symbol(
+                rxn,
+            )^C.Constraint(
+                value = -m.delta_G_reactions[Symbol(rxn)].value +
+                        dG0 +
+                        R *
+                        T *
+                        sum(
+                            m.log_metabolite_concentrations[Symbol(met_id)].value *
+                            stoich for (met_id, stoich) in zip(met_ids, stoich_coeffs)
+                        ),
+                bound = 0.0,
+            )
+    end
+
+    #= 
+    Set proton log concentration to zero so that it won't impact any
+    calculations (biothermodynamics assumption). Also set water concentrations
+    to zero (aqueous condition assumptions). How true is "aqueous conditions"? 
+    Debatable... 
+    =#
+    log_met_conc_zero = [Symbol.(proton_ids); Symbol.(water_ids)]
+    for met in keys(m.log_metabolite_concentrations)
```

Comment:
this unnecessarily checks the log_met_conc_zero many times (and that check is slow), better swap the cycle order (because check for key membership is easy). Also might be more easy to just throw an error if people don't have the IDs correct.